File: binary-tree-paths.cpp
//Runtime: 6 ms
class Solution {
public:

    void solve(vector<string>&res, string s, TreeNode* root)
    {

        if(!root->left && !root->right)
            res.push_back(s);

        if(root->left)
            solve(res, s+"->"+to_string(root->left->val), root->left);

        if(root->right)
            solve(res, s+"->"+to_string(root->right->val), root->right);
    }

    vector<string> binaryTreePaths(TreeNode* root) {
        vector<string> res;
        string s;
        if(!root) return res;
        solve(res, s+to_string(root->val), root);
        return res;
    }
};

File: binary-tree-paths.py
# Runtime: 47 ms
class Solution(object):

    def solve(self, root, t, res):
        if not root:
            return

        t += "->" + str(root.val)
        if not root.left and not root.right:
            res.append(t[2:])
            return

        self.solve(root.left, t, res)
        self.solve(root.right, t, res)

    def binaryTreePaths(self, root):
        res = []
        self.solve(root, "", res)
        return res
